src/fiber/fiber.ml

```diff
@@ -1,243 +1,455 @@
+    | l -> Parallel_iter (l, f, k)
+
+  let parallel_map l ~f k =
+    match l with
+    | [] -> k []
+    | [ x ] -> f x (fun x -> k [ x ])
+    | _ -> Parallel_map (l, f, k)
+
+  let reraise_all l _k =
+    match l with
+    | [] -> Never ()
+    | [ exn ] -> Exn_with_backtrace.reraise exn
+    | _ -> Reraise_all l
+
+  let end_of_fiber = End_of_fiber ()
+
+  let never _k = Never ()
+
+  module Ivar = struct
+    type 'a t = 'a ivar
+
+    let create () = { state = Empty }
 
-  let current = ref (create ())
+    let read t k = Read_ivar (t, k)
 
-  module K = struct
-    type 'a t = 'a k
+    let fill t x k = Fill_ivar (t, x, k)
 
-    let create run = { run; ctx = !current }
+    let peek t k =
+      k
+        (match t.state with
+        | Empty
+        | Empty_with_readers _ ->
+          None
+        | Full x -> Some x)
+  end
+
+  module Var = struct
+    include Univ_map.Key
+
+    let get var k = Get_var (var, k)
+
+    let get_exn var =
+      map (get var) ~f:(function
+        | None -> failwith "Fiber.Var.get_exn"
+        | Some value -> value)
 
-    let enqueue k x =
-      assert (k.ctx.jobs == !current.jobs);
-      Queue.push k.ctx.jobs (Job (k, x))
+    let set var x f k = Set_var (var, x, f, k)
+
+    let unset var f k = Unset_var (var, f, k)
+
+    let create () = create ~name:"var" (fun _ -> Dyn.string "var")
   end
 
-  let add_refs n =
-    let t = !current in
-    match t.on_release with
-    | Do_nothing -> ()
-    | Exec r -> r.ref_count <- r.ref_count + n
-
-  let rec forward_exn_with_backtrace t exn =
-    match t.on_error with
-    | None -> Exn_with_backtrace.reraise exn
-    | Some { ctx; run } -> (
-      current := ctx;
-      try run exn with
-      | exn -> forward_error exn)
-
-  and forward_error exn =
-    let exn = Exn_with_backtrace.capture exn in
-    forward_exn_with_backtrace !current exn
-
-  let deref t =
-    match t.on_release with
-    | Do_nothing -> ()
-    | Exec r -> (
+  type fill = Fill : 'a ivar * 'a -> fill
+
+  module Jobs = struct
+    type t =
+      | Empty
+      | Job : context * ('a -> effect) * 'a -> t
+      | Concat : t * t -> t
+
+    let concat a b =
+      match (a, b) with
+      | Empty, x
+      | x, Empty ->
+        x
+      | _ -> Concat (a, b)
+
+    let rec enqueue_readers (readers : (_, [ `Empty ]) ivar_state) x jobs =
+      match readers with
+      | Empty -> jobs
+      | Empty_with_readers (ctx, k, readers) ->
+        enqueue_readers readers x (Concat (Job (ctx, k, x), jobs))
+
+    let fill_ivar ivar x jobs =
+      match ivar.state with
+      | Full _ -> failwith "Fiber.Ivar.fill"
+      | (Empty | Empty_with_readers _) as readers ->
+        ivar.state <- Full x;
+        enqueue_readers readers x jobs
+
+    let rec exec_fills fills acc =
+      match fills with
+      | [] -> acc
+      | Fill (ivar, x) :: fills ->
+        let acc = fill_ivar ivar x acc in
+        exec_fills fills acc
+
+    let exec_fills fills = exec_fills (List.rev fills) Empty
```

Comment:
Why do we reverse? Just to make sure `Ivar.read`'s are called in the same order?

Reply:
Yes, that's the idea. I'll add a comment about it